### PR TITLE
Feature/htr

### PR DESCRIPTION
**Pull request by ferranbt**
This PR includes support for two things:
- Byte slices (i.e. [32]byte)
- Hashing of the structs with the function HashTreeRoot.

### CODE REVIEW

File: spectests/structs_test.go [hunk 1]
```diff
@@ -228,6 +230,7 @@ func TestSpecMinimal(t *testing.T) {
 			t.Fatalf("name %s not found", name)
 		}
 
+		fmt.Println(f)
```
Comment (prestonvanloon) [inline spectests/structs_test.go:233]: t.Log or remove
```suggestion
		t.Log(f)
```

File: spectests/structs_test.go [hunk 2]
```diff
@@ -248,6 +251,7 @@ func TestSpecMainnet(t *testing.T) {
...
```
Comment (prestonvanloon) [inline spectests/structs_test.go:254]: t.Log or remove
```suggestion
		t.Log(f)
```

File: sszgen/main.go [hunk 3]
```diff
@@ -315,56 +315,28 @@ func (e *env) generateEncodings() (map[string]string, error) {
 	return outs, nil
 }
 
-var errorFunctions = map[string]string{
```
Comment (prestonvanloon) [inline sszgen/main.go:318]:
```
100644 → 100755
```

Should this stay with file permission 644? Why does it need to be executable?

File: sszgen/marshal.go [hunk 4]
```diff
@@ -11,16 +11,15 @@ import (
 func (e *env) marshal(name string, v *Value) string {
 	tmpl := `// MarshalSSZ ssz marshals the {{.name}} object
 	func (:: *{{.name}}) MarshalSSZ() ([]byte, error) {
-		buf := make([]byte, ::.SizeSSZ())
-		return ::.MarshalSSZTo(buf[:0])
+		return ssz.MarshalSSZ(::)
```
Comment (prestonvanloon) [inline sszgen/marshal.go:14]: Same here. Shouldn't this file be 644?

**Comment by prestonvanloon** [comment]
Please add a HashTreeRoot interface so we can do checks like this:
```go
if v, ok := object.(ssz.HashTreeRooter); ok {
  return v.HashTreeRoot()
}
```

File: sszgen/main.go [hunk 5]
```diff
@@ -315,56 +315,28 @@ func (e *env) generateEncodings() (map[string]string, error) {
...
-}
-
 func (e *env) print(first bool, order []string) (string, error, bool) {
 	tmpl := `// Code generated by fastssz. DO NOT EDIT.
 	package {{.package}}
 	
 	import (
-		{{ if .errorFuncs }}"fmt"
-		{{ end }}
-		ssz "github.com/ferranbt/fastssz" {{ if .imports }}{{ range $value := .imports }}
```
Comment (prestonvanloon) [inline sszgen/main.go:336]: This breaks imports. Please revert.

File: sszgen/main.go (hunk 3 above)
Comment (ferranbt) [reply sszgen/main.go:318]: Good catch. Not sure why but the permissions of the files have changed.

File: sszgen/main.go (hunk 5 above)
Comment (ferranbt) [reply sszgen/main.go:336]: Good catch.

**Comment by prestonvanloon** [comment]
Would it be possible to change the return type to `[32]byte, error`? 
This would be inline with go-ssz.HashTreeRoot and other custom implementations in Prysm.

**Comment by ferranbt** [comment]
Sure, no problem

**Comment by prestonvanloon** [comment]
Here's some benchmarks I ran to compare this against go-ssz.HashTreeRoot. Please don't interpret the execution time to be how long HTR takes, this benchmark was reading files from disk, unmarshalling, marshalling, and some byte comparisions against spec tests.

[test.log](https://github.com/ferranbt/fastssz/files/4974680/test.log)

**Comment by ferranbt** [comment]
The 'htr_off' benchmarks are done with go-ssz or with the custom functions in prysm? I am not getting the same results on my benchmarks. I used [this](https://github.com/ferranbt/fastssz/blob/feature/htr/spectests/structs_test.go#L342) Htr benchmark for both fastssz and go-ssz and got:

```
goos: linux
goarch: amd64
pkg: github.com/ferranbt/fastssz/spectests
BenchmarkHashTreeRootFast
BenchmarkHashTreeRootFast-8    	    8347	    138159 ns/op	       0 B/op	       0 allocs/op
BenchmarkHashTreeRootGoSSZ
BenchmarkHashTreeRootGoSSZ-8   	     724	   1440258 ns/op	  264128 B/op	    6638 allocs/op
PASS
ok  	github.com/ferranbt/fastssz/spectests	3.120s
```

where 'BenchmarkHashTreeRootGoSSZ' is

```
import (
	oldSSZ "github.com/prysmaticlabs/go-ssz"
)

func BenchmarkHashTreeRootGoSSZ(b *testing.B) {
	obj := new(BeaconBlock)
	readValidGenericSSZ(nil, benchmarkTestCase, obj)

	b.ReportAllocs()
	b.ResetTimer()

	for i := 0; i < b.N; i++ {
		goSSZ.HashTreeRoot(obj)
	}
}
```

The numbers seem accurate for fastssz. The implementation reuses all the memory and the main factor that accounts for most of the time consumed is the sha hashing.

**Comment by prestonvanloon** [comment]
> The 'htr_off' benchmarks are done with go-ssz or with the custom functions in prysm? 

They are done with go-ssz, except for the BeaconState which uses its custom HTR function. As I mentioned before, don't interpret these times as HTR only, it does a lot more than only HTR. Here's the benchmark on a working branch: https://github.com/prysmaticlabs/prysm/blob/d89a4af4637b5fe00e3ad242d55f70ba8de3be39/proto/testing/ssz_static_mainnet_test.go#L11

**Pull request merged**